### PR TITLE
Meghana urerfbranchfix test

### PR DESCRIPTION
**Pull request by falkben**
This continues the work to fix urerf in PR #306

@megh1241 made the fix and I added a python class and pytest to ensure it is working.

I decided to make a new class, and copied most of the code, instead of adding new functionality to the existing rerf class/module.  This could be considered bad design -- let me know what you think.  Initially thought that we'd be using a very different design for doing urerf, but now that it's done, I think we could shoehorn it into our existing rerf class, just overwriting defaults when necessary.

Also updated docs to include urerf and added a demo notebook.

### CODE REVIEW

File: Python/rerf/urerf.py [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+        implementation of Random Forest. "S-RerF" is structured RerF,
+        combining multiple features together in random patches.
+        See Tomita et al. (2016) [#Tomita]_ for further details.
+    n_estimators : int, optional (default: 100)
+        Number of trees in forest.
+    max_depth : int or None, optional (default=None)
+        The maximum depth of the tree. If None, then nodes are expanded
+        until all leaves are pure or until all leaves contain less than
+        min_samples_split samples.
+    min_samples_split : int, optional (default: 1)
```
Comment (MrAE) [inline Python/rerf/urerf.py]: I thought we had determined default of 1 resulted in poor performance.
However, its current default is "auto"

File: packedForest/src/forestTypes/unsupervisedForests/baseUnprocessedNodeUnsupervised.h [hunk 2]
```diff
@@ -22,7 +22,7 @@ namespace fp{
 				std::vector<T> featureHolder;
 				
 			public:
-				baseUnprocessedNodeUnsupervised(int numObsForRoot):  parentID(0), depth(0), isLeftNode(true){
+				baseUnprocessedNodeUnsupervised(int numObsForRoot):  parentID(0), depth(0), isLeftNode(true){
```
Comment (MrAE) [inline packedForest/src/forestTypes/unsupervisedForests/baseUnprocessedNodeUnsupervised.h]: trailing whitespace.

File: packedForest/src/forestTypes/unsupervisedForests/urerf/fpURerFBase.h [hunk 3]
```diff
@@ -109,30 +109,33 @@ namespace fp {
...
-                        inline std::vector<int> predictClassPost(std::vector<T>& observation){
-				std::cout<<"Not implemented for unsupervised forests\n";
-                                return {};
-                        }
 
-			inline float testForest(){
-				std::cout<<"Not implemented for unsupervised forests\n";
+			inline float reportOOB()
+			{
 				return 0;
```
Comment (MrAE) [inline packedForest/src/forestTypes/unsupervisedForests/urerf/fpURerFBase.h:126]: If not used, should this return -1?

File: packedForest/src/forestTypes/unsupervisedForests/urf/fpURFBase.h [hunk 4]
```diff
@@ -129,24 +129,25 @@ namespace fp {
...
+			{
+				std::cout << "Not defined for unsupervised random forests. \n";
+				return {};
+			}
 
-			inline float reportOOB(){
-				std::cout<<"Not defined for unsupervised random forests. \n";
+			inline float reportOOB()
+			{
 				return 0;
```
Comment (MrAE) [inline packedForest/src/forestTypes/unsupervisedForests/urf/fpURFBase.h:145]: If not used, should this return -1?

**Review by MrAE — CHANGES_REQUESTED** [review]
Cool. A few minor comments

**Comment by MrAE** [comment]
I'm fine with it being its own class. If it lasts long enough, we can shoehorn it in later if absolutely necessary.

File: packedForest/src/forestTypes/unsupervisedForests/urerf/fpURerFBase.h (hunk 3 above)
Comment (falkben) [reply packedForest/src/forestTypes/unsupervisedForests/urerf/fpURerFBase.h:126]: @megh1241

**Review by MrAE — APPROVED** [review]

**Pull request merged**